src/Storages/KeyDescriptionMonotonicityChecker.h

```diff
@@ -0,0 +1,38 @@
+#pragma once
+
+#include <Storages/KeyDescription.h>
+#include <Interpreters/Context_fwd.h>
+#include <Storages/MergeTree/RPNBuilder.h>
+#include <Functions/IFunction.h>
+
+namespace DB
+{
+
+/*
+ * Code adapted from Storages/MergeTree/KeyCondition.
+ *
+ * Algorithm outline:
+ * 1. Navigate through partition key AST and extract RPN functions as well as the partition key column
+ * 2. Loops through RPN functions and builds a list of Functions/IFunction
+ * 3. Loops through IFunction list, grabs monotonicity info on column Range and computes intermediate Range for next iterations.
+ * */
+class KeyDescriptionMonotonicityChecker
+{
+public:
+    static IFunction::Monotonicity getMonotonicityInfo(const KeyDescription & key_description, const Range & range, ContextPtr context);
```

Comment:
It looks a bit similar to `MonotonicityCheckVisitor`, can we reuse it?

Reply:
IIRC, I failed to use it because of the API. It requires parameters like `const std::unordered_set<String> & group_by_function_hashes;` and `const TablesWithColumns & tables;` which I did not know where to get from.

Plus, I see some small differences in the logic:

1. `KeyDescriptionMonotonicityChecker` supports functions with two arguments as long as one of them is constant, `MonotonicityCheckVisitor` [does not](https://github.com/ClickHouse/ClickHouse/blob/master/src/Interpreters/MonotonicityCheckVisitor.h#L101).
2. `MonotonicityCheckVisitor` has some `canOptimize` logic based on group by and aggregate function which I did not quite understand, but does not seem valid for this use case.

I'll look into it again, if you have suggestions, please let me know.

Reply:
> It requires parameters like const std::unordered_set<String> & group_by_function_hashes; and const TablesWithColumns & tables;

Let's read the code of `MonotonicityCheckVisitor` and try to understand the meaning of these parameters:
 - `group_by_function_hashes` is only used [here](https://github.com/ClickHouse/ClickHouse/blob/ba4072f049bbf22a9b40f088527fe461ad36d910/src/Interpreters/MonotonicityCheckVisitor.h#L41-L45). Looks like it's used to avoid incorrect optimizations of SELECT queries. You don't need this restriction, so you can simply pass empty `group_by_function_hashes`
 - `tables` is only used [here](https://github.com/ClickHouse/ClickHouse/blob/ba4072f049bbf22a9b40f088527fe461ad36d910/src/Interpreters/MonotonicityCheckVisitor.h#L67-L83), and `extractIdentifierAndType` is only used [here](https://github.com/ClickHouse/ClickHouse/blob/ba4072f049bbf22a9b40f088527fe461ad36d910/src/Interpreters/MonotonicityCheckVisitor.h#L120-L125). Looks like it's needed to determine the types of the function arguments. We have some ASTIdentifier in AST, and it's a column name. This code takes the column's type from the table definition. I hope it gives a hint about how to use this parameter.

> MonotonicityCheckVisitor has some canOptimize logic based on group by and aggregate function which I did not quite understand

Again, let's read the code and try to understand. `canOptimize` checks two things: `group_by_function_hashes.count(key)` and `ast_function.is_window_function`. Yes, it's related to some SELECT queries optimizations. But if you pass empty `group_by_function_hashes` and don't have window functions, then it will just return true. So maybe we shouldn't care about `canOptimize` in this case and consider it as a part of a more generic algorithm. 


I understand that you don't want to read some "legacy code" and it may be easier to reinvent the wheel. But I don't really want to review and verify the new code that reinvents it. And if we merge it, we will have to support this code.

I didn't take a deep dive into `MonotonicityCheckVisitor` either, so maybe I'm mistaken, and it cannot be reused, but at first glance it looks like it can (maybe with a few small extensions/adjustments).

Even if it cannot be reused, it may give some useful hints about checking monotonicity (for example, it's not clear how this code checks monotonicity without the RPN-stuff you use, maybe it's possible to simplify `KeyDescriptionMonotonicityChecker`?)

Reply:
> > It requires parameters like const std::unordered_set & group_by_function_hashes; and const TablesWithColumns & tables;
> 
> Let's read the code of `MonotonicityCheckVisitor` and try to understand the meaning of these parameters:
> 
> * `group_by_function_hashes` is only used [here](https://github.com/ClickHouse/ClickHouse/blob/ba4072f049bbf22a9b40f088527fe461ad36d910/src/Interpreters/MonotonicityCheckVisitor.h#L41-L45). Looks like it's used to avoid incorrect optimizations of SELECT queries. You don't need this restriction, so you can simply pass empty `group_by_function_hashes`
> * `tables` is only used [here](https://github.com/ClickHouse/ClickHouse/blob/ba4072f049bbf22a9b40f088527fe461ad36d910/src/Interpreters/MonotonicityCheckVisitor.h#L67-L83), and `extractIdentifierAndType` is only used [here](https://github.com/ClickHouse/ClickHouse/blob/ba4072f049bbf22a9b40f088527fe461ad36d910/src/Interpreters/MonotonicityCheckVisitor.h#L120-L125). Looks like it's needed to determine the types of the function arguments. We have some ASTIdentifier in AST, and it's a column name. This code takes the column's type from the table definition. I hope it gives a hint about how to use this parameter.
> 
> > MonotonicityCheckVisitor has some canOptimize logic based on group by and aggregate function which I did not quite understand
> 
> Again, let's read the code and try to understand. `canOptimize` checks two things: `group_by_function_hashes.count(key)` and `ast_function.is_window_function`. Yes, it's related to some SELECT queries optimizations. But if you pass empty `group_by_function_hashes` and don't have window functions, then it will just return true. So maybe we shouldn't care about `canOptimize` in this case and consider it as a part of a more generic algorithm.
> 
> I understand that you don't want to read some "legacy code" and it may be easier to reinvent the wheel. But I don't really want to review and verify the new code that reinvents it. And if we merge it, we will have to support this code.
> 
> I didn't take a deep dive into `MonotonicityCheckVisitor` either, so maybe I'm mistaken, and it cannot be reused, but at first glance it looks like it can (maybe with a few small extensions/adjustments).
> 
> Even if it cannot be reused, it may give some useful hints about checking monotonicity (for example, it's not clear how this code checks monotonicity without the RPN-stuff you use, maybe it's possible to simplify `KeyDescriptionMonotonicityChecker`?)

I think I managed to get `MonotonicityCheckVisitor` working for this use case with some adaptations, but I realized it does not take the Range into consideration. It looks like my tests were not testing it either. Will look into it.

Reply:
Implemented, I just did not add tests in which the range is a crucial thing